multidict/_multidict_c.c

```diff
@@ -0,0 +1,1617 @@
+        if (pair_list_add(self->impl, key, value) < 0) {
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static int
+_multidict_append_items_seq(MultiDictObject *self, PyObject *arg,
+                            const char *name)
+{
+    PyObject *key   = NULL,
+             *value = NULL,
+             *item  = NULL,
+             *iter  = PyObject_GetIter(arg);
+
+    if (iter == NULL) {
+        return -1;
+    }
+
+    while ((item = PyIter_Next(iter)) != NULL) {
+        if (PyObject_Length(item) != 2) {
+            PyErr_Format(
+                PyExc_TypeError,
+                "%s takes either dict or list of (key, value) tuples",
+                name,
+                NULL
+            );
+            Py_DECREF(item);
+            Py_DECREF(iter);
+            return -1;
+        }
+
+        key   = PyTuple_GET_ITEM(item, 0);
+        value = PyTuple_GET_ITEM(item, 1);
+        pair_list_add(self->impl, key, value);
+
+        Py_DECREF(item);
+    }
+
+    Py_DECREF(iter);
+
+    if (PyErr_Occurred()) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+_multidict_list_extend(PyObject *list, PyObject *target_list)
+{
+    PyObject *item = NULL,
+             *iter = PyObject_GetIter(target_list);
+
+    if (iter == NULL) {
+        return -1;
+    }
+
+    while ((item = PyIter_Next(iter)) != NULL) {
+        if (PyList_Append(list, item) < 0) {
+            Py_DECREF(item);
+            Py_DECREF(iter);
+            return -1;
+        }
+        Py_DECREF(item);
+    }
+
+    Py_DECREF(iter);
+
+    if (PyErr_Occurred()) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+_multidict_extend_with_args(MultiDictObject *self, PyObject *arg,
+                            PyObject *kwds, const char *name, int do_add)
+{
+    PyObject *arg_items  = NULL, /* tracked by GC */
+             *kwds_items = NULL, /* new reference */
+             *impl       = NULL; /* borrowed ref */
+
+    int err = 0;
+
+    // TODO: mb can be refactored more clear
+    if (_MultiDict_Check(arg) && (kwds == NULL || kwds == Py_None)) {
+        if (MultiDict_CheckExact(arg) || CIMultiDict_CheckExact(arg)) {
+            impl = ((MultiDictObject*)arg)->impl;
+        } else if (MultiDictProxy_CheckExact(arg) || CIMultiDictProxy_CheckExact(arg)) {
+            impl = ((MultiDictProxyObject*)arg)->md->impl;
+        }
+
+        if (do_add) {
+            return _multidict_append_items(self, impl);
+        }
+
+        return _multidict_update_items(self, impl);
+    }
+
+    if (PyObject_HasAttrString(arg, "items")) {
+        if (_MultiDict_Check(arg)) {
+            arg_items = multidict_items((MultiDictObject*)arg);
+        } else {
+            arg_items = PyMapping_Items(arg);
+        }
+        if (arg_items == NULL) {
+            return -1;
+        }
+    } else {
+        arg_items = arg;
+        Py_INCREF(arg_items);
+    }
+
+    if (kwds && kwds != Py_None) {
```

Comment:
How `kwds` can be `None`?
I thought only `NULL` or dict can happen.

Reply:
Please call [`PyArg_ValidateKeywordArguments()`](https://docs.python.org/3/c-api/arg.html#c.PyArg_ValidateKeywordArguments) before usage.